lib/common/format.cc

```diff
@@ -0,0 +1,616 @@
+  case 'I': token_.set_kind(TokenKind::I); break;
+  case 'L': token_.set_kind(TokenKind::L); break;
+  case 'O': token_.set_kind(TokenKind::O); break;
+  case 'P': token_.set_kind(TokenKind::P); break;
+  case 'R':
+    switch (LookAheadChar()) {
+    case 'C': Advance(TokenKind::RC); break;
+    case 'D': Advance(TokenKind::RD); break;
+    case 'N': Advance(TokenKind::RN); break;
+    case 'P': Advance(TokenKind::RP); break;
+    case 'U': Advance(TokenKind::RU); break;
+    case 'Z': Advance(TokenKind::RZ); break;
+    default: token_.set_kind(TokenKind::None); break;
+    }
+    break;
+  case 'S':
+    switch (LookAheadChar()) {
+    case 'P': Advance(TokenKind::SP); break;
+    case 'S': Advance(TokenKind::SS); break;
+    default: token_.set_kind(TokenKind::S); break;
+    }
+    break;
+  case 'T':
+    switch (LookAheadChar()) {
+    case 'L': Advance(TokenKind::TL); break;
+    case 'R': Advance(TokenKind::TR); break;
+    default: token_.set_kind(TokenKind::T); break;
+    }
+    break;
+  case 'X': token_.set_kind(TokenKind::X); break;
+  case 'Z': token_.set_kind(TokenKind::Z); break;
+  case '-':
+  case '+': token_.set_kind(TokenKind::Sign); break;
+  case '/': token_.set_kind(TokenKind::Slash); break;
+  case '(': token_.set_kind(TokenKind::LParen); break;
+  case ')': token_.set_kind(TokenKind::RParen); break;
+  case '.': token_.set_kind(TokenKind::Point); break;
+  case ':': token_.set_kind(TokenKind::Colon); break;
+  case '\\': token_.set_kind(TokenKind::Backslash); break;
+  case '$': token_.set_kind(TokenKind::Dollar); break;
+  case '*':
+    token_.set_kind(LookAheadChar() == '(' ? TokenKind::Star : TokenKind::None);
+    break;
+  case ',': {
+    token_.set_kind(TokenKind::Comma);
+    CHAR laChar = LookAheadChar();
+    if (laChar == ',') {
+      Advance(TokenKind::Comma);
+      token_.set_offset(cursor_ - format_);
+      AppendError("Unexpected ',' in format expression");
+    } else if (laChar == ')') {
+      AppendError("Unexpected ',' before ')' in format expression");
+    }
+    break;
+  }
+  case '\'':
+  case '"':
+    for (++cursor_; cursor_ < end_; ++cursor_) {
+      if (*cursor_ == c) {
+        if (cursor_ >= end_ || *(cursor_ + 1) != c) {
+          token_.set_kind(TokenKind::String);
+          break;
+        }
+        ++cursor_;
+      }
+    }
+    if (stmt_ == IoStmtKind::Read) {
+      AppendError("String edit descriptor in READ format");  // 13.3.2p6
+    }
+    if (token_.kind() == TokenKind::String) {
+      break;
+    }
+    token_.set_kind(TokenKind::None);
+    cursor_ = end_;
+    token_.set_length(cursor_ - format_ - token_.offset());
+    AppendError("Unterminated string");
+    break;
+  default:
+    if (cursor_ >= end_) {
+      suppressErrorCascade_ = false;
+      AppendError("Unterminated format expression");
+    }
+    token_.set_kind(TokenKind::None);
+    break;
+  }
+
+  token_.set_length(cursor_ - format_ - token_.offset() + 1);
```

Comment:
This overrides for instance line 206 

`token_.set_length(cursor_` - format_ - token_.offset());`

Is that expected behavior?

Reply:
Yes.  I rewrote this yesterday to provide better error markers, which is what this does.  The code is now a little different, but this pattern is the same.